context.go

```diff
@@ -75,45 +93,70 @@ func (d RunAddressData) isEmpty() bool {
-func (context *Context) Run(addressData RunAddressData, timeout time.Duration) (res Result, err error) {
+// if the output of TotalTime() exceeds the value of Timeout, the function will immediately return with errors.ErrTimeout
+// The second parameter is deprecated and should be passed to NewContextWithBudget instead.
+func (context *Context) Run(addressData RunAddressData, _ time.Duration) (res Result, err error) {
 	if addressData.isEmpty() {
 		return
 	}
 
-	now := time.Now()
 	defer func() {
-		dt := time.Since(now)
-		context.totalOverallRuntimeNs.Add(uint64(dt.Nanoseconds()))
+		if err == errors.ErrTimeout {
+			context.timeoutCount.Inc()
+		}
 	}()
 
-	// At this point, the only error we can get is an error in case the top level object is a nil map, but this
-	// behaviour is expected since either persistent or ephemeral addresses are allowed to be null one at a time.
-	// In this case, EncodeAddresses will return nil contrary to Encode which will return an nil wafObject,
-	// which is what we need to send to ddwaf_run to signal that the address data is empty.
-	var persistentData *bindings.WafObject = nil
-	var ephemeralData *bindings.WafObject = nil
-	persistentEncoder := newLimitedEncoder()
-	ephemeralEncoder := newLimitedEncoder()
-	if addressData.Persistent != nil {
-		persistentData, _ = persistentEncoder.EncodeAddresses(addressData.Persistent)
+	// If the context has already timed out, we don't need to run the WAF again
+	if context.timer.SumExhausted() {
+		return Result{}, errors.ErrTimeout
 	}
 
-	if addressData.Ephemeral != nil {
-		ephemeralData, _ = ephemeralEncoder.EncodeAddresses(addressData.Ephemeral)
+	runTimer, err := context.timer.NewNode("dd.appsec.waf.run",
+		timer.WithComponents(
+			"dd.appsec.waf.encode.persistent",
+			"dd.appsec.waf.encode.ephemeral",
+			"dd.appsec.waf.duration_ext",
+			"dd.appsec.waf.duration",
+		),
+	)
+	if err != nil {
+		return Result{}, err
+	}
 
+	runTimer.Start()
+	defer func() {
+		context.stats.add("dd.appsec.waf.run", runTimer.Stop())
+		context.stats.merge(runTimer.Stats())
+	}()
+
+	persistentData, persistentEncoder, err := encodeOneAddressType(addressData.Persistent, runTimer.MustLeaf("dd.appsec.waf.encode.persistent"))
+	if err != nil {
+		return res, err
 	}
-	// The WAF releases ephemeral address data at the end of each run call, so we need not keep the Go values live beyond
+
+	// The WAF releases ephemeral address data at the max of each run call, so we need not keep the Go values live beyond
 	// that in the same way we need for persistent data. We hence use a separate encoder.
+	ephemeralData, ephemeralEncoder, err := encodeOneAddressType(addressData.Ephemeral, runTimer.MustLeaf("dd.appsec.waf.encode.ephemeral"))
+	if err != nil {
+		return res, err
+	}
 
-	// ddwaf_run cannot run concurrently and the next append write on the context state so we need a mutex
+	// ddwaf_run cannot run concurrently and the next merge write on the context state so we need a mutex
```

Comment:
That append is the one done in the `defer` above? If so, it'll happen _after_ the lock is released... which this comment would imply is not fine?